addons/account/models/account_move_line.py

```diff
@@ -626,12 +626,20 @@ def get_rate(from_currency, to_currency, company, date):
                 date=date,
             )
         for line in self:
-            line.currency_rate = get_rate(
-                from_currency=line.company_currency_id,
-                to_currency=line.currency_id,
-                company=line.company_id,
-                date=line.move_id.date or fields.Date.context_today(line),
-            )
+            if line.currency_id == line.company_id.currency_id:
+                line.currency_rate = 1
+            elif line.move_id.is_invoice(include_receipts=True):
+                line.currency_rate = get_rate(
+                    from_currency=line.company_currency_id,
+                    to_currency=line.currency_id,
+                    company=line.company_id,
+                    date=line.move_id.date or fields.Date.context_today(line),
+                )
+            else:
+                # The rate can be set manually on journal entries, hence we approximate it
+                # We don't need the dependency because we get the rate after setting the amounts
+                # in this case.
+                line.currency_rate = line.amount_currency / (line.balance or 1)
```

Comment:
I could get the exact rate instead of the approximation here if the result of `get_rate`  matches `line.amount_currency / (line.balance or 1)` 
Not sure if it's a good idea